Run the app locally and on CI using docker compose

tahb opened the pull request:
<!-- Do you need to update the changelog? -->

This change allows developers to run the application locally using Docker if they wish, by introducing Docker Compose. This is at least my preference for working with apps locally and I've found it to be generally viable as part of day-to-day dev.

We have existing scripts and documentation for starting applications _without_ Docker. These remain the same and honour the existing Scripts to Rule Them All pattern.

Using Docker in development has advantages of stronger parity. Given [we MUST always use containers to run CI and SHOULD used in live environments](https://github.com/dxw/tech-team-rfcs/blob/main/rfc-013-use-docker-to-deploy-and-run-applications-in-containers.md) using Docker locally can help identify environment shaped bugs earlier in delivery. Containerisation also rules out a whole category of issues related to our local environment, package versions etc. The script/bootstrap does attempt to deal with this but it cannot guarantee a fully compatible environment. 

Docker should always be stable given we use it for CI and live environments. Given we have the Docker work, by adding Docker compose we can leverage it for development and simpler getting started instructions (if you just want to have a play around).

## Changes in this PR

- Docker Compose support added
- CI uses Docker Compose rather than native Docker for stronger parity to development
- CI continues to use docker image caching 
- Following a change in organisation to the script directory we can now run the same scripts with Docker and as CI using variables such as `DOCKER=true` or `CI=true`
- For those that are working on the app and need to test regularly we continue to recommend testing outside of Docker. These changes only help us spin up a working server or exec onto a rails console (though a docker testing command is included)
- "Getting started" documentation moved to a standalone file to keep the README tidy and to help view a sum of the getting started options we now offer for comparison
- [We add Brakeman as part of our test suite](https://github.com/dxw/tech-team-rfcs/blob/main/rfc-024-use-brakeman.md) (this change is piggybacking on this PR and can be taken out if it becomes a problem)

## Screenshots of UI changes

Here we can see CI running with [our caching work](https://github.com/dxw/rails-template/pull/147) still in full effect (without caching this build takes ~5-7 minutes).

![Screenshot 2021-04-23 at 13 12 30](https://user-images.githubusercontent.com/912473/115871820-f3c79800-a438-11eb-8e16-dc94f57d266f.png)

tahb committed:
Commands to be used in the terminal prefixed with $

tahb committed:
Docker compose is used by CI to test the application

This codifies the knowledge of how to start a set of Docker containers into a reusable Docker Compose format.

With Docker Compose CI _and_ local development can use identical
environments and work with strong parity.

tahb committed:
Refactor how the Dockerfile bundles gems

* We now only call bundle install once, we don't have to copy arguments for `--retry` and `--jobs` across and keep them in sync
* We still use the `RAILS_ENV` as the single control for what gem groups would be installed. If running in production, this won't install dev and test. Nothing should change here but the code should be more succinct

tahb committed:
Test script runs Brakeman

[dxw use Brakeman as a static code scanner](https://github.com/dxw/tech-team-rfcs/blob/main/rfc-024-use-brakeman.md).

Run this at the end of the order as it takes a while. Usually we'd like quicker rspec feedback when working with known fixing failing tests.

Ignore `coverage/` from linting checks in case Brakeman is run without
being output to stdout

tahb committed:
CI docker image caching is less fragile

When Docker Compose builds images it will use the name of the current directory (currently Rails Template) and suffix the name of the container (test).

This will create a docker image: rails-template_test

That image name is great for the template itself but when it's used as a template a new directory name will be given, for example MyGreatApp.

CI uses `docker save` with rails-template_test for caching. If we did nothing then Docker Compose would build against an image of a different name and not be able to take advantage of the inherent advantages of building from cached docker layers.

This change removes the needs to add another TODO for the team to change this when using the template by using the Docker Compose project name argument[1]

[1] https://docs.docker.com/compose/reference/#use--p-to-specify-a-project-name

erbridge commented on doc/getting-started.md:
```diff
@@ -0,0 +1,72 @@
+
+### Prerequisites
+
+1. [Install Docker for MacOS](https://docs.docker.com/docker-for-mac/install/)
```
```suggestion
brew install --cask docker
```

erbridge commented on doc/getting-started.md:
````diff
@@ -0,0 +1,72 @@
+Start the server:
+
+```bash
+$ script/dserver
````
Can we make these still use `script/server` and `script/console`, just with arguments? For example `script/server --docker`.

tahb replied:
[As I'm hoping to discuss here](https://github.com/dxw/rails-template/pull/213#discussion_r619267748): I'm not convinced of layering in docker and non docker logic into the same scripts. I think it's cleaner to extend the standard scripts than squash them together with more variables.

tahb replied:
Resolving this thread as I'm changing the structure in new commit for review 👍

erbridge commented on doc/getting-started.md:
````diff
@@ -0,0 +1,72 @@
+Run the tests:
+
+```bash
+$ docker-compose -f docker-compose.test.yml run --rm test script/test
````
```suggestion
$ script/test --docker
```

tahb replied:
[As I'm hoping to discuss here](https://github.com/dxw/rails-template/pull/213#discussion_r619267748): I'm not convinced of layering in docker and non docker logic into the same scripts. I think it's cleaner to extend the standard scripts than squash them together with more variables.

erbridge replied:
I think either way I'd want a quicker, more memorable incantation for the tests. If we stick with the `script/dx` pattern, then I'd want a `scripts/dtest` too.

pezholio replied:
Agree with @erbridge here 👍

tahb replied:
Sure 👍

erbridge commented on .github/workflows/continuous-integration.yml:
```diff
@@ -29,27 +27,12 @@ jobs:
-            -e CI=true \
-            app:test script/test
+      - name: Build
+        run: docker-compose -f docker-compose.ci.yml -p app build
```
Can we make these use `script/build` and `script/test` and switch file by checking for the `CI` environment variable instead? I'm worried about the number of different places we have different incantations to do the same sort of thing. I'd love for us to have the standard scripts to rule them all be the entrypoints for all of this stuff, and for them to call subscripts (in subfolders, probably), depending on arguments.

tahb replied:
Do you mean have a `script/build` file that only does this docker build equivalant?

I agree only having the standard entrypoints would be nice. I'm not convinced that layering in more complexity into those scripts to support the multiple contexts is worth the trade off:

- CI
- dev (no docker)
- dev (docker) 

It wasn't the easiest thing to go through each case statement to determine what behaviour was expected to happen on CI and off CI. Mixing in docker and non docker logic across them would I fear make it very hard to work with.

erbridge replied:
I wonder if a single entry point is valuable, though. I'm not saying we necessarily end up with lots of nested casing. Instead, I'm thinking something along the lines of:

```bash
if [ $CI ]; then
  script/ci/build
elif [ $DOCKER ]; then
  script/docker/build
else
  script/no-docker/build
fi
```

The advantage of this approach is that as someone coming to the project, I only have the 5 scripts I'm familiar with to think about. You could even make running it without any arguments cause it to spit out some help text asking them to choose Docker or no Docker, to self-document. The value of the scripts to rule them all pattern is that I should be able to do almost everything I need to do in a project without needing to read the docs to find out how. First time set up? Run `script/setup`. Coming back? Run `script/update`. Tests? `script/test`. Run the thing? `script/server`. Debug the thing? `script/console`. I don't want to have to think.

erbridge replied:
(build might be a bad example here, but I hope it illustrates my general point)

pezholio replied:
[`script/cibuild`](https://github.com/github/scripts-to-rule-them-all/blob/master/script/cibuild) is the usual pattern here (though I prefer `script/ci`

tahb replied:
I think this is a great idea. It doesn't totally elimate the need to read the docs as you'd still need to know how to set `DOCKER=true` if that was your choice.

I'm going to add a new commit that reorganises the script structure in line with this 👍 I'll resolve this thread and look forward to seeing what you think on the newer commit.

Floppy commented:
Doesn't affect the PR, as this is an optional thing, but has the speed improved on Macs? I've always avoided docker in dev because of the performance problems with mounted volumes on macOS. Just wondering though, not criticising.

tahb commented:
@Floppy Yes it has improved when running as a normal server. I've used it everyday for over a year now. When you first boot it up it can take 5-10 seconds to sort itself out (presumably establishing the volumes and mounts etc) but from then on I don't notice a difference between that and a normal `rails server`.

pezholio commented on script/dconsole:
```diff
@@ -0,0 +1,8 @@
+#!/bin/sh
+
+# script/console: Launch a console for the application. Optionally pass in an
+#                 environment name to connect to a console for that environment.
```
```suggestion
# script/dconsole: Launch a console for the application alongside any other required
                   processes inside Docker containers.
```

pezholio commented on script/dserver:
```diff
@@ -0,0 +1,10 @@
+#!/bin/sh
+
+# script/server: Launch the application and any extra required processes
+#                locally.
```
```suggestion
# script/server: Launch the application and any extra required processes
#                locally in Docker containers.
```

tahb commented:
Ready for another review.

We should now be able to run the following:

```
CI=true script/server
DOCKER=true script/server
script/server

CI=true script/console
DOCKER=true script/console
script/console

CI=true script/test
DOCKER=true script/test
script/test

CI=true script/test spec/features/visitors/about_page_spec.rb
DOCKER=true script/test spec/features/visitors/about_page_spec.rb
script/test spec/features/visitors/about_page_spec.rb
```

Thanks for your feedback so far @erbridge, @pezholio.

mec commented on doc/getting-started.md:
```diff
@@ -0,0 +1,44 @@
+["Scripts To Rule Them All"](https://github.com/dxw/tech-team-rfcs/blob/main/rfc-023-use-scripts-to-rule-them-all.md)
+pattern to standardise the common commands we have to use.
+
+
```
By the final change, I think we have lost the `script/setup` command from here?

tahb replied:
Good shout 👍

mec commented on docker-entrypoint.sh:
```diff
@@ -11,6 +11,11 @@ setup_database()
   echo "ENTRYPOINT: Finished database setup."
 }
 
+# Bundle any Gemfile changes in development to avoid a long image rebuild
```
I am hoping to see some guidance around this later, the commit message talks about this need to rebuild that I am not sure about that seems to be a day-to-day need of running inside docker locally.

tahb replied:
Perhaps I can add more description. The problem we are solving here is:

1. You are running a server with Docker
2. You change the Gemfile
3. You want to `bundle` so that you can use your new gem

Before this change, you have to figure out how to update your docker image. One way to do this is to do a full rebuild with script/setup but this does everything and takes an age. By which time you've context switched for far too long. 

This change means you can do `script/server` again straight away, and it will bundle anything extra atop of the built image. This means you can have gems installed quickly and delay the full rebuild.

On reflection, this needs documentation otherwise folks won't know to use it in this way.

On further reflection, perhaps we should move `bundle` this into `script/update` so that you would do:

4. Stop the docker server
5. Run `script/update`
6. Run `script/server`

This seems much better so I think I'll change to this.

tahb replied:
Damn, no we can't call script/docker/update because it has no way to update the built image without itself calling docker-compose build web.. which perhaps it ought to do.

tahb replied:
This was removed. We instead point people to running script/update which in a Docker context will do a full rebuild and trust in its own local caching to make it fast.

mec commented on README.md:
```diff
@@ -17,80 +17,7 @@ TODO: replace README header with project name
 
 TODO: Add a summary of who the application is for and what it will do.
 
-## Getting started
```
It threw me seeing this change here and not in the commit that introduced the `getting-started.md` file - but it is not problem!

tahb replied:
Squashed

mec commented on .eslintignore:
```diff
@@ -2,3 +2,4 @@
 /tmp/
 /vendor/
 /public/assets
+/coverage
```
Are these ignore helping brakeman or the linters - the commit message has confused me a bit so I am interested to learn?

dragon-dxw replied:
There are files in coverage which the linters don't know what to do with(particularly yarn run lint:format:fix) and cause them to fail.
Also there's a few huge files which take minutes which we'd skip too.

mec commented on docker-compose.ci.yml:
```diff
@@ -5,6 +5,9 @@ services:
         RAILS_ENV: "test"
+      cache_from:
+        - app_test:latest
+    image: app_test
```
Nice! 👍

mec commented on script/ci/cibuild:
```diff
@@ -0,0 +1,8 @@
+#!/bin/sh
+
+# script/cibuild: Setup environment for CI to run tests. This is primarily
```
Some of these little path comments don't seem to be right and may not be bringing anything? I would either fix them or remove them - the comments are great, I am only speaking to the paths!?

tahb replied:
Good point. I've removed them in https://github.com/dxw/rails-template/pull/213/commits/312ce81c8ef7c8443c156b683a3986edf7767150

mec commented on doc/getting-started.md:
```diff
@@ -1,74 +1,44 @@
-["Scripts To Rule Them All"](https://github.com/dxw/tech-team-rfcs/blob/main/rfc-023-use-scripts-to-rule-them-all.md)
-pattern to standardise the common commands we have to use.
+## With Docker
 
```
This is where I might like to learn about te `bundle` and `build` business?

tahb replied:
I've removed the bundle/build problem by pointing folks to use the generic script/update which should figure it out for ppl.

mec commented on docker-compose.test.yml:
```diff
@@ -15,6 +15,7 @@ services:
     environment:
       DATABASE_URL: postgres://postgres:password@test-db:5432/app-test
       DATABASE_CLEANER_ALLOW_REMOTE_DATABASE_URL: "true"
+      AUTOMATICALLY_FIX_LINTING: "true"
```
Does this mean everywhere but CI the linters will fix without prompting?

tahb replied:
Yes, when you run script/test it'll attempt to make automatic fixes but on CI it'll cause a failure.

The general approach of automatic fixing is something originally copied from our scripts to rule them all template found here https://github.com/dxw/scripts-to-rule-them-all/blob/main/ruby/rails/script/test. I haven't tried to introduce or change this but have tried to make it work with Docker too.

mec reviewed (changes requested):
The changes here mostly look good, I've left a couple of comment/questions and I think we lost some documentation in the change over.

It was an super interesting journey for me as I am in neither the one script and installing everything or run it all in docker camps. You have opened my eyes to a benefit of having the option of docker for light touch people, perhaps on support, that I had not considered before. 👍 

The Rails tempalte always raised lots of interesting questions in my mind and I never know the right place for that discussion, but I am confident it is not here.

This works has been around a while and I would like to help get it merged as I think it is blocking other work you have!

Give me a yell if I can help with anything.

tahb commented:
@mec thanks for your review. I've made a change and/or replied to each comment with what I think 👍

mec commented:
@tahb thank you for all your work on this - I think we should keep pushing to get an approval from @erbridge or @pezholio so we can get this merged!

erbridge commented on script/dserver:
```diff
@@ -0,0 +1,10 @@
+
+docker-compose up --detach
+# TODO: Repace 'rails-template' with application name
+docker attach rails-template_web_1
```
Could you explain what's going on here in a comment next to it maybe? Why are detaching only to reattach?

tahb replied:
I've added this:

```
# INFO: To enable debug prompts (with pry or byebug) we first start the process
# in the background and then use attach to create an interactive prompt.
```

erbridge commented on Dockerfile:
```diff
@@ -29,14 +29,10 @@ COPY Gemfile.lock $INSTALL_PATH/Gemfile.lock
+ENV BUNDLE_GEM_GROUPS=$RAILS_ENV
+RUN bundle config set no-cache "true"
+RUN bundle config set with $BUNDLE_GEM_GROUPS
+RUN bundle install --no-binstubs --retry=10 --jobs=4
```
Why `--no-binstubs`?

tahb replied:
I thinking that as were are installing gems into a pristine environment we might not need the extra stubs. I wondered if we could shave off a bit of build time by skipping it.

I checked to see if we can run bin/rails and bin/rake from inside the container and we can. Thinking about it now though I guess I wasn't taking account of any gems that want to add their own binstubs.

erbridge commented on doc/getting-started.md:
```diff
@@ -1,74 +1,49 @@
+### Usage
 
-Start a console:
+Set `DOCKER` before running any of the above commands.
```
```suggestion
Set `DOCKER=true` before running any of the above commands.
```

erbridge commented on doc/getting-started.md:
```diff
@@ -1,74 +1,49 @@
+### Usage
 
-Start a console:
+Set `DOCKER` before running any of the above commands.
```
Is it worth suggesting something like [direnv](https://github.com/direnv/direnv) as a tool for setting this as a project default? I also wonder if it would make sense to give it a more specific name (or alias) eg `PREFER_DOCKER_FOR_DXW_RAILS` so you can set it in your *shrc file as a global default?

tahb replied:
I agree that `PREFER_DOCKER_FOR_DXW_RAILS` is much safer to use in rc files than `DOCKER`. It's a long one for users to write when using one-off commands. I'll add in the suggestion to use global and/or direnv.

tahb replied:
I added a new switch and stopped pointing uses towards using `docker=true`.

erbridge commented on script/console:
```diff
@@ -1,25 +1,20 @@
+# script/console: Launch a console for the application. Use variables to control
+#                 how you want to provision your local environment.
+#                 Eg. `DOCKER=true script/console`
+#                 Or, add this as a global variable to your shell so it's set
+#                 automatically.
```
FIXME: I'd worry about collisions from setting `DOCKER=true` globally. What else will now try to use Docker because of it?

tahb replied:
Good point

erbridge commented on script/console:
```diff
@@ -1,25 +1,20 @@
 
 set -e
 
-cd "$(dirname "$0")/.."
```
FIXME: What happens if you run the script from somewhere other than the root of the project? This line was there to ensure we're in the right place.

tahb replied:
Do I need something else in place before calling scripts from anywhere works? For me I need to provide a path to it from where I am:

```
cd app
script/console
=> zsh: command not found: script/console
```

tahb replied:
I understand how this works now, will add in.

erbridge commented on script/no-docker/bootstrap:
```diff
@@ -5,8 +5,6 @@
 set -e
 
-cd "$(dirname "$0")/.."
-
```
FIXME: Again, shouldn't we keep this to make sure we're executing the rest of this script from the correct location?

```suggestion
cd "$(dirname "$0")/../.."
```

erbridge commented on script/no-docker/setup:
```diff
@@ -0,0 +1,58 @@
+set -e
+
+if [ -z "$CI" ] && [ -n "$DB_DROP_RESULT" ]; then
+  printf "\\nDatabase drop failed. Continue anyway? [y/N] "
+  read -r
```
You can use the `-p` argument to do this with `read` directly, fyi.

tahb replied:
As I only moved this existing code across into a new file I'm inclined to leave this one as it was to keep my changeset from growing anymore.

erbridge commented on script/no-docker/update:
```diff
@@ -4,10 +4,8 @@
 
 set -e
 
-cd "$(dirname "$0")/.."
```
FIXME: Unless I'm missing something, my comments on these deletions apply everywhere, including the new files that should have them added.

erbridge commented on doc/getting-started.md:
````diff
@@ -53,3 +53,17 @@ Set `DOCKER` before running any of the above commands.
+Add the following line to the bottom of your `~/.bash-profile` (or `~/.zshrc` if
+you use Z Shell):
+
+```
+export DOCKER=true
+```
````
FIXME: As I've said elsewhere, a global environment variable with such a generic name is dangerous. Who knows what else watches for it? We'd be better off giving a more specific name (or an alias that's more specific).

I also strongly recommend something like direnv so you can do this on a project by project basis if you want.

erbridge commented on doc/getting-started.md:
````diff
@@ -53,3 +53,17 @@ Set `DOCKER` before running any of the above commands.
 ```
+
+To set Docker as the default you can add this variable into your environment.
+Add the following line to the bottom of your `~/.bash-profile` (or `~/.zshrc` if
````
Should this be the profile or the rc?

erbridge commented on script/run_all_tests:
```diff
@@ -1,7 +1,8 @@
 
-# script/test: Run the test suite for the application. Optionally pass in a path
-#              to an individual test file to run a single test.
+# script/run_all_tests: Run the test suite for the application. Optionally pass
+#                       in a path to an individual test file to run a single
+#                       test.
```
Shouldn't this have its path removed, too?

erbridge commented on script/no-docker/update:
```diff
@@ -1,6 +1,6 @@
 #!/bin/sh
 
-# script/update: Update application to run for its current checkout.
+# Update application to run for its current checkout.
```
I'd find these a lot more useful absorbed into the original commits where possible instead of in their own commit here.

erbridge reviewed (commented):
This approach looks sound to me, but I do wonder about the churn in the PR. We introduce a lot of things that we then modify or remove. Could we have reordered the commits and resulted in a cleaner history instead?

There are a few functional issues in this review (marked FIXME) and some suggestions. Don't let me block the merge once you've fixed (or confirmed as not an issue) the functional issues.

tahb committed:
Scripts reorganised to better support: ci, docker and no-docker

* We continue to have a single interface for running common commands: server, console and test.
* If we want to use these commands with Docker we call the same script with DOCKER=true set
* If we want to use these commands to debug CI we can also do that
* Splitting the script structure into multiple files requires us to change the way we call `shellcheck`. If we don't do this, it doesn't test nested files and it errors when given a folder.
* Maintain a single place where we define how every environment/build context should test the application in `run_all_tests`. We remove the distinction between CI and not and add in a new variable called `AUTOMATICALLY_FIX_LINTING` which better describe the distinction.

tahb committed:
Use a switch to specify if docker should be used

We still use the original DOCKER variable under the hood as the trigger for how the scripts should operate.

The way we set this value has now changed to allow for 2 mechanisms.

1. For regular users we allow a descriptive `PREFER_DOCKER_FOR_DXW_RAILS` to be set in the users environment (whether globally or on a project level). This brings a descriptive name that should not conflict as the very general `DOCKER` might have done.
2. We're also still able to have short hand ways of using these scripts infrequently by using a switch. This was more preferable than writing `DOCKER=true` before each command since users might try to automate this by adding it into their rc files - which may cause those same conflcits we're trying to avoid

tahb commented:
@erbridge I've hopefully addressed all the FIXME comments, I added in the switch for `--docker` etc and cleaned up the git history (which I was keeping in case the new advised approach didn't work out).

erbridge commented on doc/getting-started.md:
```diff
@@ -45,3 +45,33 @@ Running the tests regularly should be done without Docker as it is much faster.
+
+### Occassional use
+
+Use the `--docker` or `-d` switch:
```
[optional] Tiny objection to `-d` as a shortening. My brain goes "as a daemon" for that, so I'd probably expect that to run it in a background process or something.

erbridge commented on doc/getting-started.md:
```diff
@@ -45,3 +45,33 @@ Running the tests regularly should be done without Docker as it is much faster.
+You can set `PREFER_DOCKER_FOR_DXW_RAILS=1` in your env to set this preference
+automatically.
+
+- Option 1:
+  - Install and use
+    [direnv to set an environment variable on a per codebase basis](https://direnv.net/)
+- Add `PREFER_DOCKER_FOR_DXW_RAILS=1` to your `.envrc`
+
+  - Load change with `direnv allow .`
+- Option 2:
+- Set `PREFER_DOCKER_FOR_DXW_RAILS=1` in your global `.zshrc` or `.bashrc`
+
+  - Load your change with `source .zshrc`
```
The line breaks and indentation in here make it kind of hard to follow (and I think don't match what you intended). Does it need some attention to make sure things are indented correctly. If Prettier formatted it in this way for you, it's because that's how a markdown parser would read it.

tahb replied:
It was auto-formatting that I missed. I'm just going to turn it off at this point.

erbridge replied:
The autoformatting probably surfaced a syntax error, btw. Prettier works by parsing what you pass to it and then rewriting that, so this is what it understood.

erbridge commented on script/bootstrap:
```diff
@@ -8,7 +8,22 @@ set -e
 
-if [ "$DOCKER" ]; then
+echo "==> Loading command line options…"
+export DOCKER=${PREFER_DOCKER_FOR_DXW_RAILS:-0}
+for arg in "$@"
+do
+  case $arg in
+    -d|--docker)
+      export DOCKER=1
+      ;;
+    --no-docker)
+      export DOCKER=0
+      ;;
+  esac
+  shift
+done
```
We don't need `export`s in here and probably don't want to leak them into the environment we run subprocesses of this script in, which is what `export` will do. Using `export` is like setting a global variable. If you do `VARIABLE=value` without `export` you're making it local to the currently executed script and anything it `source`s.

You also only need to `export` a variable once. Once it's exported, reassigning it will reassign the thing that's exported.

tahb replied:
Yeah fair. I was trying to have this run in another script which ran in a sub shell (if that's a thing). Trying to export it outwards into a different context. Failed and copy pasted it back into each file without re-reading.

erbridge replied:
Oh, you might also not want the `shift` unless you mean to clear all of the arguments passed in.

erbridge commented on script/console:
```diff
@@ -1,25 +1,23 @@
-#                 environment name to connect to a console for that environment.
+# script/console: Launch a console for the application. Use variables to control
+#                 how you want to provision your local environment.
+#                 Eg. `DOCKER=true script/console`
+#                 Or, add this as a global variable to your shell so it's set
+#                 automatically.
```
This documentation is out of date now.

erbridge reviewed (approved):
Some minor fixes and it's good to go, I think. I'm assuming that the script changes are essentially renames, so haven't looked at them again in any detail.

tahb committed:
Fix running of specific tests through script/test

`--pattern` is not the conventional way we tell rspec what files to run. As per the documentation [1] we can simply pass files or paths for multiple files in as a parameter.

Using the `--pattern` control presents the user with a load file error.

```
$ bundle exec rspec --pattern spec
An error occurred while loading ./spec.
Failure/Error: return load_without_bootsnap(resolved, wrap)

LoadError:
  cannot load such file -- /Users/x/sites/dxw/rails-template/spec
```

[1] https://rspec.info/documentation/4.0/rspec-rails/#Running_specs
